daml/ContingentClaims/Lifecycle.daml

```diff
@@ -3,83 +3,120 @@
+import DA.Action ((>=>))
+import DA.Traversable (sequence)
 import Prelude hiding (sequence, mapA)
 
 type C a = Claim Observation Date a
 type F a = ClaimF Observation Date a
 
 deriving instance Eq a => Eq (C a)
 
--- | Returned from a lifecycle operation.
+-- | Returned from a `settle` operation.
 data Result a = Result with
   remaining : C a
     -- ^ the tree after lifecycled branches have been pruned.
   pending : [(Decimal, a)]
     -- ^ quantity/asset pairs requiring settlement.
 
--- | Lifecycle a Claim. Parameters:
--- 1. Function taking key and date, used to lookup an observable on the ledger. `m` will typically be `Update`. It is parametrised so it can be run in a `Script`.
--- 2. Function used when branching in an `Or`. Must return one of the arguments.
--- 3. A set of `Claim`s.
--- 4. A function which, given a date, lifecycles the third argument.
-lifecycle : (Eq a, CanAbort m)
+-- | Collect claims falling due into a list, and return the tree with those nodes pruned.
+-- `m` will typically be `Update`. It is parametrised so it can be run in a `Script`.
+-- The first argument is used to lookup the value of any `Observables`.
+-- Results in a function taking today's date, and returning the pruned tree + pending settlements.
+settle : (Eq a, CanAbort m)
   => (Text -> Date -> m Decimal)
-  -> (C a -> C a -> C a)
   -> C a
   -> Date -> m (Result a)
-lifecycle spot choose
+settle spot
   = runKleisli
   . fmap (uncurry Result)
   . flip evalStateT 1.0
   . runWriterT
-  . apoCataM (pruneZeros') (lifecycle' spot choose)
+  . apoCataM pruneZeros' settleLazily'
+    where
+      settleLazily' = (lift . lift <$> lazily' spot)
+                      >=> (fmap distM (fmap (settle' spot . embed)) . sequence)
 
-type Seed m a = WriterT [(Decimal, a)] (StateT Decimal (Kleisli m Date)) (F a (Either (C a) (C a)))
+-- | HIDE The monad stack we're using.
+type M m a = WriterT [(Decimal, a)] (StateT Decimal (Kleisli m Date)) (F a (Either (C a) (C a)))
+--           \----------------------- monad ------------------------/  \-------- value --------/
 
-lifecycle' : (Eq a, Monad m, CanAbort m)
+-- | HIDE Helper function for sequencing algebras.
+distM : Monad m
+      => Either (C a) (WriterT [(Decimal, a)] (StateT Decimal (Kleisli m Date)) (F a (C a)))
+      -> M m a
+distM (Left c) = pure (Left <$> (project c))
+distM (Right m) = fmap Right <$> m
+
+-- | Evaluate observables and skip branches for which predicates don't hold.
+-- This is useless onl it's own. Composed with other functions it adds laziness.
+lazily' : (Eq a, Monad m)
   => (Text -> Date -> m Decimal)
-  -> (C a -> C a -> C a)
-  -> C a -> Seed m a
+  -> C a -> Kleisli m Date (F a (Either (C a) (C a)))
+lazily' spot (When obs c) = do
+  predicate <- eval spot obs
+  pure if predicate then WhenF obs (Right c) else WhenF obs (Left c)
+lazily' spot (Cond obs c c') = do
+  predicate <- eval spot obs
+  if predicate then lazily' spot c else lazily' spot c'
```

Comment:
Remove these `lazily'`, as @leonidr-da pointed out, explicit recursion is confusing here.

Reply:
This can't be done without changing this to a futumorphism. That adds yet another type (`Free`) that is used to 'unfold' the tree several levels ahead. In this case, it's safe to call the op recursively, as we are only 'consuming' new nodes. I'd like to change this, but I'm afraid that doing so would add even more complexity at the type-level, so I'll leave as-is for now.